<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.6: http://docutils.sourceforge.net/" />
<title>Working with BLAST outputs</title>
<style type="text/css">

BODY {
	margin: 0;
	padding: 10px;
	font-family: Georgia, serif;
	font-size: 16px;
	color: #000;
}

PRE {
	font-family: Fixedsys, Monaco, Courier;
	font-size: 10pt;
	padding: 2px 0 2px 7pt;
	border-left: 5px solid #d2f2f1;
	margin-left: 40px;
	color: #003f3f;
	background: #faffff;
	overflow: auto;
	width: 90%;
 }

CODE, TT {
	font-family: Fixedsys, Monaco, Courier;
	font-size: 10pt;
	color: #8F1D00;
 }

H1.title {
	margin-top: 0px;
	font-size: 21px;
	border: none;
	text-align: left;
}

H1 {
	font-size: 19px;
	border-bottom: 1px solid #aaa;
	margin-top: 20px;
}

H2 {
	font-size: 18px;
	margin-left: 20px;
	margin-top: 20px;
}

H3 {
	font-size: 16px;
	margin-left: 20px;
}

H4 {
	font-size: 16px;
	margin-left: 20px;
	font-weight: normal;
}

P {
	text-indent: 20px;
}

A, A:hover, A:visited {
	color: #000;
	text-decoration: none;
	text-transformation: none;
	border-bottom: 1px dashed #aaa;	
}

H2 > A, H2 > A:hover, H2 > A:visited,
H3 > A, H3 > A:hover, H3 > A:visited,
H4 > A, H4 > A:hover, H4 > A:visited,
DIV.contents * A, DIV.contents * A:visited {
	border-bottom: none;
}

DIV.note {
	border: 1px solid #7099C5;
	background: #f4ffff;
	padding: 0;
	margin-left: 20px;
	margin-right: 20px;
	-webkit-border-radius: 8px;
	-moz-border-radius: 8px;
}

DIV.note > P.first {
	font-weight: bold;
	display: inline;
	padding: 4px 0 0 4px;
}
</style>
</head>
<body>
<div class="document" id="working-with-blast-outputs">
<h1 class="title">Working with BLAST outputs</h1>

<p>MetagenomeDB considers BLAST results as relationship between sequences; namely, between each query and its hits. Once BLAST results are imported in the database, you can ask for all hits that align against a given query sequence, or the opposite. The alignment itself (coordinates and sequence of the HSP) and the scores (E-value, percentage of identity) are stored and can be used to filter down your queries. Information about the BLAST run, such as the program used, its version, and the parameters are stored as well.</p>
<div class="section" id="importing-blast-outputs">
<h1>Importing BLAST outputs</h1>
<p>The <tt class="docutils literal"><span class="pre">mdb-import-BLAST-alignments</span></tt> tool is provided to import NCBI BLAST outputs:</p>
<pre class="literal-block">
$ mdb-import-BLAST-alignments --help
Usage: mdb-import-BLAST-alignments [options]

Part of the MetagenomeDB toolkit. Imports XML-formatted NCBI BLAST alignments
into the database.

Options:
  -h, --help            show this help message and exit
  -v, --verbose
  --dry-run

  Input:
    -i FILENAME, --input=FILENAME
                        XML-formatted output of a NCBI BLAST sequence
                        alignment (mandatory).
    -Q STRING, --query-collection=STRING
                        Name of the collection the query sequences belong to
                        (mandatory).
    -H STRING, --hit-collection=STRING
                        Name of the collection the hit sequences belong to
                        (optional). If not provided, the hit sequences are
                        assumed to be external to the database, and only a
                        summary of those hits will be stored: hit identifer,
                        description and E-value.
    --date=YEAR MONTH DAY
                        Date of the BLAST run (optional). By default, creation
                        date of the input file.
    --query-id-getter=PYTHON CODE
                        Python code to reformat query identifers (optional);
                        '%' will be replaced by the query identifier. Default:
                        %.split()[0]
    --hit-id-getter=PYTHON CODE
                        Python code to reformat hit identifers (optional); '%'
                        will be replaced by the hit identifier. Default:
                        %.split()[0]
    --no-check          If set, bypass the query and hit sequences identifier
                        check (not recommended).

  Input filtering:
    --max-E-value=FLOAT
                        If set, filter out all hits with a E-value above the
                        provided cut-off.
    --min-identity=INTEGER
                        If set, filter out all hits with a percentage of
                        identity below the provided cut-off.
    --max-hits=INTEGER  If set, keep only the first '--max-hits' hits for each
                        query.
    --ignore-alignment  If set, will not store information about the sequence
                        alignment (HSP coordinates and sequences).

  Connection:
    --host=HOSTNAME     Host name or IP address of the MongoDB server
                        (optional). Default: localhost
    --port=INTEGER      Port of the MongoDB server (optional). Default: 27017
    --db=STRING         Name of the database in the MongoDB server (optional).
                        Default: 'MetagenomeDB'
    --user=STRING       User for the MongoDB server connection (optional).
                        Default: ''
    --password=STRING   Password for the MongoDB server connection (optional).
                        Default: ''
</pre>
<p>This tool accepts XML-formatted NCBI BLAST outputs only. Those outputs can be concatenated (i.e., one XML file resulting from the concatenation of several BLAST outputs). Here is a review of the most important parameters:</p>
<div class="section" id="input-sequences">
<h2>Input sequences</h2>
<p>There are two situations this tool can handle:</p>
<ul class="simple">
<li>Alignments among sequences that are already in the database. In this situation, both the query and hit sequences are objects present in the database. Options <tt class="docutils literal"><span class="pre">-Q</span></tt> (<tt class="docutils literal"><span class="pre">--query-collection</span></tt>) and <tt class="docutils literal"><span class="pre">-H</span></tt> (<tt class="docutils literal"><span class="pre">--hits-collection</span></tt>) are requested to known the name of the collections containing the query and hit sequences, respectively.</li>
<li>Alignments against an external database. In this situation, the query sequences are objects present in the database while the hits are external to it (e.g., NR). Option <tt class="docutils literal"><span class="pre">-Q</span></tt> (<tt class="docutils literal"><span class="pre">--query-collection</span></tt>) must be provided.</li>
</ul>
</div>
<div class="section" id="sequence-identifiers">
<h2>Sequence identifiers</h2>
<p>The optional <tt class="docutils literal"><span class="pre">--query-id-getter</span></tt> and <tt class="docutils literal"><span class="pre">--hit-id-getter</span></tt> options can be used to modify query and hit identifiers on-the-fly, respectively. This is useful if you expect the BLAST XML output to contain sequence identifiers with some additional, unwanted characters.</p>
<p>For example, considering the following XML output:</p>
<pre class="literal-block">
...
&lt;Iteration&gt;
  &lt;Iteration_iter-num&gt;1&lt;/Iteration_iter-num&gt;
  &lt;Iteration_query-ID&gt;1&lt;/Iteration_query-ID&gt;
  &lt;Iteration_query-def&gt;CH0704v-contig00010 length=3963   numreads=678&lt;/Iteration_query-def&gt;
  &lt;Iteration_query-len&gt;3963&lt;/Iteration_query-len&gt;
  &lt;Iteration_hits&gt;
    ...
</pre>
<p>The query identifier that <tt class="docutils literal"><span class="pre">mdb-import-BLAST-alignments</span></tt> will consider is the whole string &quot;CH0704v-contig00010 length=3963   numreads=678&quot;. If only the first element (&quot;CH0704v-contig00010&quot;) is needed, you can use the following Python code for the <tt class="docutils literal"><span class="pre">--query-id-getter</span></tt> option:</p>
<pre class="literal-block">
$ mdb-import-BLAST-alignments --query-id-getter &quot;%.split()[0]&quot;
</pre>
<p>Any '%' character in the string you provide will be replaced by the value of the query identifier. In this example, the short Python code used will split the original string (resulting in the list &quot;CH0704v-contig00010&quot;, &quot;length=3963&quot;, &quot;numreads=678&quot;) and select the first element. The same approach applies for <tt class="docutils literal"><span class="pre">--hit-id-getter</span></tt>.</p>
</div>
<div class="section" id="hits-filtering">
<h2>Hits filtering</h2>
<p>The optional options <tt class="docutils literal"><span class="pre">--max-E-value</span></tt>, <tt class="docutils literal"><span class="pre">--min-identity</span></tt> and <tt class="docutils literal"><span class="pre">--max-hits</span></tt> can be used to ignore some of the hits contained in the BLAST output.</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">--max-E-value</span></tt> will filter out any hit with a E-value above a provided cut-off</li>
<li><tt class="docutils literal"><span class="pre">--min-identity</span></tt> will filter out any hit with a percentage of identity below a provided cut-off</li>
<li><tt class="docutils literal"><span class="pre">--max-hits</span></tt> will filter out any hit with position greater than a cut-off in the BLAST output</li>
</ul>
</div>
<div class="section" id="connection">
<h2>Connection</h2>
<p>The optional <tt class="docutils literal"><span class="pre">--host</span></tt>, <tt class="docutils literal"><span class="pre">--port</span></tt>, <tt class="docutils literal"><span class="pre">--db</span></tt>, <tt class="docutils literal"><span class="pre">--user</span></tt> and <tt class="docutils literal"><span class="pre">--password</span></tt> options are common to all MetagenomeDB tools and can be used to bypass the default server connection information found in <tt class="docutils literal"><span class="pre">~/.MetagenomeDB</span></tt>.</p>
</div>
</div>
<div class="section" id="querying-blast-results">
<h1>Querying BLAST results</h1>
<p>As introduced at the top of the document, <tt class="docutils literal"><span class="pre">mdb-import-BLAST-alignments</span></tt> convert BLAST outputs into relationship between query and hit sequences. In practice, those relationships are annotated with information about the alignment, such as various scores and the algorithm used.</p>
<p>Let us consider an alignment between two sequences A (query) and B (hit) in the database. After importing the BLAST output you could perform the following queries:</p>
<pre class="literal-block">
# We select the sequence 'A'
A = mdb.Sequence.find_one({&quot;name&quot;: &quot;A&quot;})

# We ask for all relationships of type 'part-of'
# this sequence could have with other sequences
for hit in A.list_related_sequences(mdb.Direction.INGOING, {&quot;type&quot;: &quot;part-of&quot;}):
        print hit
</pre>
</div>
</div>
</body>
</html>
